### PR TITLE
[TNH_002]host-signup-login

### PR DESCRIPTION
**Pull request by JakeGiangan**
[Commands]
1. rails db:migrate:reset
2. rails g devise:install
3. rails s
4. access via localhost:3000

[Pre-conditions]

1.  create a DB in MySQL Workbench named travellers_nomadic_hut with username: root password: password at port 3306

[Expected Output]

1.  New User (Host) can sign up, login and logout

[Notes]

1. Only to see if you logged in/logged out/ created a new account successfully or failed, i temporarily added a flash message that will appear at the middle of the header (its design is still temporary)
2. The drop down at the top will change when the user is logged in or not
3. Currently it only covers the Host Accounts, Guests/Renter accounts will be made in a different PR
4. made some minor adjustments for the homepage to give it some space
5. footer is always fixed at the bottom of the screen

[Screenshots]
image

### CODE REVIEW

**Review by charlotteds — CHANGES_REQUESTED** [review]
Notes:

For the Sign Up and Authentication in this project, Devise will be used.
References:
https://github.com/heartcombo/devise
https://docs.google.com/presentation/d/1QT-5iy2XIrDotpFOIW-Odyduk8T5QbJ4SMbbenqGY38/edit?usp=sharing

You may use the slides for the Expert Course as guide (Please don't edit any of these):
https://drive.google.com/drive/u/1/folders/1GvLgQ2zsdhl5ZXXK4VNkq1BeVHRApCX6

File: app/assets/javascripts/users.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (charlotteds) [inline app/assets/javascripts/users.coffee]: Delete unused file.

File: test/models/user_test.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (charlotteds) [inline test/models/user_test.rb]: Delete unused file.

File: test/controllers/users_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (charlotteds) [inline test/controllers/users_controller_test.rb]: Delete unused file.

**Review by charlotteds — COMMENTED** [review]
Fix indentation for all files:
- HTML (`.html.erb`): 4 spaces
- RUBY (`.rb`): 2 spaces

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -1,16 +1,25 @@
 <!DOCTYPE html>
```
Comment (charlotteds) [inline app/views/layouts/application.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/confirmations/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
```
Comment (charlotteds) [inline app/views/users/confirmations/new.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/mailer/email_changed.html.erb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
```
Comment (charlotteds) [inline app/views/users/mailer/email_changed.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/passwords/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
```
Comment (charlotteds) [inline app/views/users/passwords/edit.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/passwords/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<h2>Forgot your password?</h2>
```
Comment (charlotteds) [inline app/views/users/passwords/new.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/registrations/edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,43 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (charlotteds) [inline app/views/users/registrations/edit.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/registrations/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,57 @@
+<% provide(:title, "Sign up as a host") %>
```
Comment (charlotteds) [inline app/views/users/registrations/new.html.erb]: Fix indentation. Use 4 spaces.

File: app/views/users/sessions/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,35 @@
+<% provide(:title, "Log in") %>
```
Comment (charlotteds) [inline app/views/users/sessions/new.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/shared/_error_messages.html.erb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<% if resource.errors.any? %>
```
Comment (charlotteds) [inline app/views/users/shared/_error_messages.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/shared/_links.html.erb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
```
Comment (charlotteds) [inline app/views/users/shared/_links.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/views/users/unlocks/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend unlock instructions</h2>
```
Comment (charlotteds) [inline app/views/users/unlocks/new.html.erb:1]: Fix indentation. Use 4 spaces.

File: app/controllers/application_controller.rb [hunk 15]
```diff
@@ -1,2 +1,13 @@
...
+  protect_from_forgery with: :exception, prepend: true
+
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+
+    def configure_permitted_parameters
+      devise_parameter_sanitizer.permit(:sign_up) { |u| u.permit(:first_name, :last_name, :birthdate, :languages, :contact_number, :account_type, :email, :password)}
+
+      devise_parameter_sanitizer.permit(:account_update) { |u| u.permit(:first_name, :last_name, :birthdate, :languages, :contact_number, :account_type, :email, :password, :current_password)}
```
Comment (charlotteds) [inline app/controllers/application_controller.rb:11]: Remove `email`, `password`, and `current_password`. These are already supported by Devise.

Comment (JakeGiangan) [reply app/controllers/application_controller.rb:11]: this PR isn't using Devise just yet, the Devise implementation was implemented in PR #4 including the indentions changes as well

File: app/views/users/sessions/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        </div>
+        <div class="col-lg-8 offset-lg-2">
+            <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
+                <div class="row fieldRowSpace" >
+                <div class="col">
+                <%= f.text_field :email, class: "roundedFormField" ,placeholder: "Email Address", style: "min-width: 300px" %>
+                </div>
+                <div class="col">
+                <%= f.password_field :password, class: "roundedFormField" ,placeholder: "Password", style: "min-width: 300px" %>
+                </div>
```
Comment (charlotteds) [inline app/views/users/sessions/new.html.erb]: Indent fields between the `<div>` `</div>`

File: app/views/users/registrations/new.html.erb (hunk 10 above)
Comment (charlotteds) [reply app/views/users/registrations/new.html.erb]: unresolved

File: app/controllers/application_controller.rb (hunk 15 above)
Comment (charlotteds) [reply app/controllers/application_controller.rb:11]: It seems like you have added Devise to this PR already with your commit: "converted to use Devise Gem".
Please make this PR work with Devise before adding any changes to the others.

**Review by charlotteds — APPROVED** [review]
Can Sign up, Login, and Logout successfully.

**Review by christanplaza — APPROVED** [review]
Was able to register, login and logout properly.

Approved.

**Pull request merged**